### PR TITLE
Pulling "default" literal feature into master branch

### PR DESCRIPTION
**Pull request by jcouv**
Implements https://github.com/dotnet/csharplang/blob/master/proposals/target-typed-default.md

@cston @dotnet/roslyn-compiler My understanding is that we intend to do code reviews when merging features back into master, so I won't merge without approvals.

### CODE REVIEW

**Comment by 333fred** [comment]
I hosed the debug integration test while investigating failures

File: src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs [hunk 1]
```diff
@@ -500,21 +500,28 @@ private bool GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder builder, BoundE
...
+                }
             }
 
             if ((object)collectionExprType == null) // There's no way to enumerate something without a type.
             {
-                // The null literal was caught above, so anything else with a null type is a method group or anonymous function
-                diagnostics.Add(ErrorCode.ERR_AnonMethGrpInForEach, _syntax.Expression.Location, collectionExpr.Display);
-                // CONSIDER: dev10 also reports ERR_ForEachMissingMember (i.e. failed pattern match).
-
+                if (collectionExpr.Kind == BoundKind.DefaultExpression && (object)collectionExpr.Type == null)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs]: Second condition has been checked above.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 2]
```diff
@@ -517,7 +517,9 @@
     <Field Name="GetFieldFromHandle" Type="MethodSymbol" Null="allow"/>
   </Node>
 
-  <Node Name="BoundDefaultOperator" Base="BoundExpression">
+  <Node Name="BoundDefaultExpression" Base="BoundExpression">
+    <!-- Type is null in the case of a default literal, and non-null in the case of a fully-spelled out default operator. -->
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: Perhaps `"and non-null for default<T>"`

File: src/Compilers/CSharp/Portable/BoundTree/Constructors.cs [hunk 3]
```diff
@@ -523,10 +523,15 @@ public static BoundBlock SynthesizedNoLocals(SyntaxNode syntax, params BoundStat
...
     {
-        public BoundDefaultOperator(SyntaxNode syntax, TypeSymbol type)
-            : this(syntax, type.GetDefaultValue(), type)
+        public BoundDefaultExpression(SyntaxNode syntax, TypeSymbol type, bool hasErrors = false)
+            : this(syntax, type.GetDefaultValue(), type, hasErrors)
+        {
+        }
+
+        public BoundDefaultExpression(SyntaxNode syntax)
+          : this(syntax, constantValueOpt: null, type: null, hasErrors: false)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/Constructors.cs]: `LiteralExpressionSyntax syntax`?

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 4]
```diff
@@ -5044,4 +5053,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_CompilerAndLanguageVersion" xml:space="preserve">
     <value>Compiler version: '{0}'. Language version: {1}.</value>
   </data>
+  <data name="WRN_DefaultInSwitch" xml:space="preserve">
+    <value>Did you mean to use the default switch label (`default:`) rather than `case default:`? If you really mean to use the default literal, consider `case (default):` or another literal (`case 0:` or `case null:`) as appropriate.</value>
+  </data>
+  <data name="WRN_DefaultInSwitch_Title" xml:space="preserve">
+    <value>Did you mean to use the default switch label (`default:`) rather than `case default:`? If you really mean to use the default literal, consider `case (default):` or another literal (`case 0:` or `case null:`) as appropriate.</value>
+  </data>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: How are `_Title` resource strings used? (This is a long title.)

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 5]
```diff
@@ -1472,5 +1472,9 @@ internal enum ErrorCode
         ERR_FeatureNotAvailableInVersion7_1 = 8302,
         ERR_LanguageVersionCannotHaveLeadingZeroes = 8303,
         ERR_CompilerAndLanguageVersion = 8304,
+
+        ERR_BadDynamicMethodArgDefaultLiteral = 9000,
+        ERR_DefaultLiteralNotValid = 9001,
+        WRN_DefaultInSwitch = 9002,
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: Should the `ErrorCode` gap be smaller before merging into master?

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 4 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: I don't know. We have a test that requires warnings to have matching resource strings, and those generally match the original except they don't have substitutions.

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs (hunk 5 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: I will compact once in master.

**Comment by gafter** [comment]
@cston Where would I find the test plan for this?

**Comment by cston** [comment]
@gafter https://github.com/dotnet/roslyn/pull/18595

File: src/Compilers/CSharp/Portable/BoundTree/Constructors.cs (hunk 3 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/Constructors.cs]: No, this bound node can represent both `default` (a literal) and `default(T)` (not a literal).

**Comment by jcouv** [comment]
I'll create a new PR (so that I can fix CreateCompilationWithMscorlib issue in one shot).

**Pull request closed without merging**